File: rust/bron_kerbosch/src/bron_kerbosch9.rs
//! Bron-Kerbosch algorithm with pivot and degeneracy ordering, optimized

use bron_kerbosch_degeneracy::degeneracy_order_smart;
use bron_kerbosch_pivot::{visit, PivotChoice};
use graph::UndirectedGraph;
use pile::Pile;
use reporter::{Clique, Reporter};
use util::intersect;

extern crate crossbeam;
use std::collections::HashSet;
use std::sync::mpsc;
use std::sync::mpsc::Sender;

struct SendingReporter {
    tx: Sender<Clique>,
}

impl Reporter for SendingReporter {
    fn record(&mut self, clique: Clique) {
        self.tx.send(clique).unwrap();
    }
}

pub fn explore(graph: &UndirectedGraph, reporter: &mut Reporter) {
    let mut candidates = graph.connected_nodes();
    debug_assert_eq!(
        degeneracy_order_smart(graph, &candidates).collect::<HashSet<_>>(),
        candidates
    );
    let mut excluded = HashSet::with_capacity(candidates.len());

    let (tx, rx) = mpsc::channel();
    crossbeam::thread::scope(|scope| {
        for v in degeneracy_order_smart(graph, &candidates) {
            let neighbours = graph.adjacencies(v);
            debug_assert!(!neighbours.is_empty());
            candidates.remove(&v);
            let neighbouring_candidates = intersect(&neighbours, &candidates).cloned().collect();
            let neighbouring_excluded = intersect(&neighbours, &excluded).cloned().collect();
            excluded.insert(v);
            let thread_tx = tx.clone();
            scope.spawn(move |_| {
                visit(
                    graph,
                    &mut SendingReporter { tx: thread_tx },
                    PivotChoice::MaxDegree,
                    PivotChoice::MaxDegree,
                    neighbouring_candidates,
                    neighbouring_excluded,
                    Pile::Cons(&Pile::Empty, v),
                );
            });
        }
        drop(tx);
        while let Ok(clique) = rx.recv() {
            reporter.record(clique);
        }
    })
    .unwrap();
}

File: python3/bron_kerbosch9.py
# coding: utf-8

from bron_kerbosch_pivot import pick_max_degree, visit
from graph import UndirectedGraph, Vertex
from reporter import Reporter
from dataclasses import dataclass, field
import queue
from typing import List, Set


def explore(graph: UndirectedGraph, reporter: Reporter):
    '''Bron-Kerbosch algorithm with pivot and degeneracy ordering,
    optimized'''
    reporter.inc_count()
    candidates = graph.connected_nodes()
    excluded: Set[Vertex] = set()
    assert candidates == set(
        degeneracy_order_smart(graph=graph, candidates=candidates))
    for v in degeneracy_order_smart(graph=graph, candidates=candidates):
        neighbours = graph.adjacencies[v]
        assert neighbours
        candidates.remove(v)
        visit(
            graph=graph,
            reporter=reporter,
            initial_pivot_choice=pick_max_degree,
            further_pivot_choice=pick_max_degree,
            candidates=candidates.intersection(neighbours),
            excluded=excluded.intersection(neighbours),
            clique=[v])
        excluded.add(v)


@dataclass(order=True)
class PrioritizedItem:
    priority: int
    node: Vertex = field(compare=False)


def degeneracy_order_smart(graph: UndirectedGraph, candidates: Set[Vertex]):
    priority_per_node = [-2] * graph.order
    q: queue.PriorityQueue = queue.PriorityQueue()
    for c in candidates:
        p = graph.degree(c)
        priority_per_node[c] = p
        q.put(PrioritizedItem(priority=p, node=c))

    for _ in range(len(candidates)):
        i = q.get(block=False).node
        while priority_per_node[i] == -1:
            i = q.get(block=False).node
        assert priority_per_node[i] >= 0
        priority_per_node[i] = -1
        yield i
        for v in graph.adjacencies[i]:
            p = priority_per_node[v]
            if p != -1:
                assert p > 0
                priority_per_node[v] = p - 1
                q.put(PrioritizedItem(priority=p - 1, node=v), block=False)
